### ISSUE TITLE
Teacher unsupervised loss mask multiplied outside sum

### ISSUE THREAD

**Issue by dagleaves** [kekmodel/MPL-pytorch#7]
First, I have to say this is such great work, and it has been a great help in understanding the MPL code not coming from a TF background. 

https://github.com/kekmodel/MPL-pytorch/blob/c66b4d3c45bc15c4be915b82eda50044f8cb3c04/main.py#L180-L182

I would not say that I am confident that this is an issue, but it is an inconsistency that I would like to bring to your attention in case it is. In the original code, they multiply the mask inside the sum as shown below, but yours multiplies outside. Does this not change the values from what they would be if multiplied inside? Or does it not matter?

![image](https://user-images.githubusercontent.com/68761152/109895963-24bbf580-7c5e-11eb-866a-9aba6de1e5bb.png)

**Comment by kekmodel** [comment]
`mask`'s shape is `batch_size x 1` and `-(soft_pseudo_label * torch.log_softmax(t_logits_us, dim=-1)).sum(dim=-1)`'s shape is also `batch_size x 1`. And `torch.mean` = `tf.reduce_sum / batch_size`. So I think it's the right formula.

**Comment by dagleaves** [comment]
Right! My comment wasn't on the sizes though. I was mentioning the order of operations. In the original code, they multiply the -cross-entropy['u'] by the mask and then sum the result. 

Yours takes the negative sum of cross_entropy['u'] (equivalent to soft_pseudo_label * log_softmax(t_logits_us, dim=-1) and multiplies that by mask. 

My question was does multiplying the mask outside of the sum not change the result?

**Comment by kekmodel** [comment]
Since the `mask` consists only of 1 and 0, I think the order is irrelevant.

**Comment by dagleaves** [comment]
Ahh, excellent! That explains it, thank you!
